implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_worker.rs

```diff
@@ -0,0 +1,344 @@
+                )
+                .await?
+        };
+
+        // if we reached the final state we can make a pair of encryptor/decryptor
+        if let Some(final_state) = self.state_machine.get_handshake_results() {
+            // start the encryptor worker and return the decryptor
+            self.decryptor_handler = Some(self.finalize(context, final_state).await?);
+            self.callback_sender.send(()).await?;
+        };
+
+        Ok(())
+    }
+}
+
+impl HandshakeWorker {
+    /// Create a new HandshakeWorker with a role of either INITIATOR or RESPONDER
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) async fn create(
+        context: &Context,
+        secure_channels: Arc<SecureChannels>,
+        addresses: Addresses,
+        identifier: IdentityIdentifier,
+        trust_policy: Arc<dyn TrustPolicy>,
+        decryptor_outgoing_access_control: Arc<dyn OutgoingAccessControl>,
+        credentials: Vec<Credential>,
+        trust_context: Option<TrustContext>,
+        remote_route: Option<Route>,
+        timeout: Option<Duration>,
+        role: Role,
+    ) -> Result<()> {
+        let vault = to_xx_vault(secure_channels.vault());
+        let identities = secure_channels.identities();
+        let state_machine: Box<dyn StateMachine> = if role.is_initiator() {
+            Box::new(
+                InitiatorStateMachine::new(
+                    vault,
+                    identities,
+                    identifier.clone(),
+                    credentials.clone(),
+                    trust_policy.clone(),
+                    trust_context.clone(),
+                )
+                .await?,
+            )
+        } else {
+            Box::new(
+                ResponderStateMachine::new(
+                    vault,
+                    identities,
+                    identifier.clone(),
+                    credentials.clone(),
+                    trust_policy.clone(),
+                    trust_context.clone(),
+                )
+                .await?,
+            )
+        };
+
+        let (mut callback_waiter, callback_sender) = ockam_node::callback::new_callback();
+
+        let worker = Self {
+            secure_channels,
+            callback_sender,
+            state_machine,
+            identifier: identifier.clone(),
+            role: role.clone(),
+            remote_route: remote_route.clone(),
+            addresses: addresses.clone(),
+            decryptor_handler: None,
+        };
+
+        WorkerBuilder::new(worker)
+            .with_mailboxes(Self::create_mailboxes(
+                &addresses,
+                decryptor_outgoing_access_control,
+            ))
+            .start(context)
+            .await?;
+
+        let decryptor_remote = addresses.decryptor_remote.clone();
+        debug!(
+            "Starting SecureChannel {} at remote: {}",
+            role.clone(),
+            &decryptor_remote
+        );
+        trace!(
```

Comment:
I think this should be either removed or moved down to where the channel is already initialized